File: app/src/main/java/org/thoughtcrime/securesms/blocked/BlockedUsersAdapter.java
package org.thoughtcrime.securesms.blocked;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.util.Consumer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import org.thoughtcrime.securesms.R;
import org.thoughtcrime.securesms.components.AvatarImageView;
import org.thoughtcrime.securesms.recipients.Recipient;
import org.thoughtcrime.securesms.util.DateUtils;

import java.util.Locale;
import java.util.Objects;

final class BlockedUsersAdapter extends ListAdapter<Recipient, BlockedUsersAdapter.ViewHolder> {

  private final RecipientClickedListener recipientClickedListener;

  BlockedUsersAdapter(@NonNull RecipientClickedListener recipientClickedListener) {
    super(new RecipientDiffCallback());

    this.recipientClickedListener = recipientClickedListener;
  }

  @Override
  public @NonNull ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.blocked_users_adapter_item, parent, false),
                          position -> recipientClickedListener.onRecipientClicked(Objects.requireNonNull(getItem(position))));
  }

  @Override
  public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
    holder.bind(Objects.requireNonNull(getItem(position)));
  }

  static final class ViewHolder extends RecyclerView.ViewHolder {

    private final AvatarImageView avatar;
    private final TextView        displayName;
    private final TextView        username;
    private final TextView        blockedAt;

    public ViewHolder(@NonNull View itemView, Consumer<Integer> clickConsumer) {
      super(itemView);

      this.avatar           = itemView.findViewById(R.id.avatar);
      this.displayName      = itemView.findViewById(R.id.display_name);
      this.username         = itemView.findViewById(R.id.username);
      this.blockedAt        = itemView.findViewById(R.id.blocked_at);

      itemView.setOnClickListener(unused -> {
        if (getAdapterPosition() != RecyclerView.NO_POSITION) {
          clickConsumer.accept(getAdapterPosition());
        }
      });
    }

    public void bind(@NonNull Recipient recipient) {
      avatar.setAvatar(recipient);
      displayName.setText(recipient.getDisplayName(itemView.getContext()));

      String identifier = recipient.getUsername().orElse(null);

      if (identifier != null) {
        username.setText(identifier);
        username.setVisibility(View.VISIBLE);
      } else {
        username.setVisibility(View.GONE);
      }

      long blockedAtTimestamp = recipient.getBlockedAt();
      if (blockedAtTimestamp > 0) {
        String time = DateUtils.formatDateWithYear(Locale.getDefault(), blockedAtTimestamp);
        blockedAt.setText(itemView.getContext().getString(R.string.BlockedUsersActivity__blocked_on_s, time));
        blockedAt.setVisibility(View.VISIBLE);
      } else {
        blockedAt.setVisibility(View.GONE);
      }
    }
  }

  private static final class RecipientDiffCallback extends DiffUtil.ItemCallback<Recipient> {

    @Override
    public boolean areItemsTheSame(@NonNull Recipient oldItem, @NonNull Recipient newItem) {
      return oldItem.equals(newItem);
    }

    @Override
    public boolean areContentsTheSame(@NonNull Recipient oldItem, @NonNull Recipient newItem) {
      return oldItem.equals(newItem);
    }
  }

  interface RecipientClickedListener {
    void onRecipientClicked(@NonNull Recipient recipient);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/blocked/BlockedUsersAdapter.kt
/*
 * Copyright 2025 Signal Messenger, LLC
 * SPDX-License-Identifier: AGPL-3.0-only
 */

package org.thoughtcrime.securesms.blocked

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.util.Consumer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import org.thoughtcrime.securesms.R
import org.thoughtcrime.securesms.components.AvatarImageView
import org.thoughtcrime.securesms.recipients.Recipient
import java.util.Objects
import kotlin.jvm.optionals.getOrNull

class BlockedUsersAdapter(private val recipientClickedListener:RecipientClickedListener) :
  ListAdapter<Recipient, BlockedUsersAdapter.ViewHolder>(RecipientDiffCallback()) {
  
  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
    return ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.blocked_users_adapter_item, parent, false)) {
      position: Int ->
      recipientClickedListener.onRecipientClicked(getItem(position))
    }
  }

  override fun onBindViewHolder(holder: ViewHolder, position: Int) {
    holder.bind(Objects.requireNonNull(getItem(position)))
  }

  
  class ViewHolder(itemView: View, clickConsumer: Consumer<Int>) : RecyclerView.ViewHolder(itemView) {
    private val avatar: AvatarImageView = itemView.findViewById(R.id.avatar)
    private val displayName: TextView = itemView.findViewById(R.id.display_name)
    private val username: TextView = itemView.findViewById(R.id.username)

    init {
      itemView.setOnClickListener { _ ->
        if (bindingAdapterPosition != RecyclerView.NO_POSITION) {
          clickConsumer.accept(bindingAdapterPosition)
        }
      }
    }

    fun bind(recipient: Recipient) {
      avatar.setAvatar(recipient)
      displayName.text = recipient.getDisplayName(itemView.context)

      val identifier = recipient.username.getOrNull()
      username.apply{
        text = identifier ?: ""
        visibility = identifier?.let { View.VISIBLE } ?: View.GONE
      }
    }
  }

  class RecipientDiffCallback : DiffUtil.ItemCallback<Recipient>() {
    override fun areItemsTheSame(oldItem: Recipient, newItem: Recipient): Boolean {
      return oldItem == newItem
    }

    override fun areContentsTheSame(oldItem: Recipient, newItem: Recipient): Boolean {
      return oldItem == newItem
    }
  }

  fun interface RecipientClickedListener {
    fun onRecipientClicked(recipient: Recipient)
  }
}



